File: docs/js/pool-complex.js
import { secureRandomInt, fmtChips } from "./core.js";
import { isNetPositive } from "./hotel.js";
import { recordConsumption, POOL_CONSUMABLE_IDS } from "./intoxication-effects.js";

/** @typedef {{ ok: boolean, message: string, unlock?: string }} PoolResult */

export const POOL_ZONES = {
  wave_pool: {
    id: "wave_pool",
    label: "Wave Pool",
    description: "Eleven acres of artificial surf. The wave machine hums like a slot reel.",
  },
  hot_tubs: {
    id: "hot_tubs",
    label: "Hot Tubs",
    description: "Steaming circles of gossip, chlorine, and questionable life choices.",
  },
  cabanas: {
    id: "cabanas",
    label: "Private Cabanas",
    description: "Rope-off luxury. Bottle service implied. Dignity optional.",
  },
  shark_reef: {
    id: "shark_reef",
    label: "Shark Reef Aquarium",
    description: "Sand tiger sharks glide behind acrylic. Children press faces to the glass. You relate.",
  },
  beach_club: {
    id: "beach_club",
    label: "Topless Beach Club",
    description: "21+ European-style pool deck. Cover charge enforced. Confidence recommended.",
    requiresAge: 21,
  },
  beach_rave: {
    id: "beach_rave",
    label: "Beach Rave",
    description: "Neon after dark. DJ, glow sticks, and the wave pool lit like a fever dream.",
  },
};

export const SHARK_SPECIES = {
  sand_tiger: { id: "sand_tiger", label: "Sand tiger shark", points: 1 },
  green_sea_turtle: { id: "green_sea_turtle", label: "Green sea turtle", points: 1 },
  sawfish: { id: "sawfish", label: "Largetooth sawfish", points: 2 },
  golden_ray: { id: "golden_ray", label: "Golden cownose ray", points: 1 },
  jellyfish: { id: "jellyfish", label: "Pacific sea nettle", points: 1 },
};

const WAVE_TIMING = ["Jump early", "Ride the crest", "Bail late"];

export const POOL_EVENTS = {
  first_splash: {
    id: "first_splash",
    label: "First Splash",
    narrative: "The wave catches you perfectly. Chlorinated triumph.",
    requires: { zones: ["wave_pool"] },
  },
  steam_and_chips: {
    id: "steam_and_chips",
    label: "Steam & Chips",
    narrative: "Hot tub gossip confirms your blackjack run. The bubbles applaud.",
    requires: { zones: ["hot_tubs"], netPositive: true },
  },
  cabana_king: {
    id: "cabana_king",
    label: "Cabana King",
    narrative: "Your cabana. Your rules. The ice bucket respects you.",
    requires: { flags: ["cabana_booked"] },
  },
  reef_photographer: {
    id: "reef_photographer",
    label: "Reef Photographer",
    narrative: "Five species captured. The gift shop sells your enthusiasm back to you.",
    requires: { sharkPhotos: 5 },
  },
  beach_club_initiate: {
    id: "beach_club_initiate",
    label: "Beach Club Initiate",
    narrative: "You survived the rope line. The sun deck nods approvingly.",
    requires: { zones: ["beach_club"] },
  },
  rave_til_dawn: {
    id: "rave_til_dawn",
    label: "Rave Til Dawn",
    narrative: "Glow sticks, bass drops, and a wave pool that shouldn't exist at 2 AM.",
    requires: { zones: ["beach_rave"], flags: ["rave_danced"] },
  },
  eleven_acres: {
    id: "eleven_acres",
    label: "Eleven Acres Conquered",
    narrative: "Every zone in the complex. The lifeguard salutes. You need a nap.",
    requires: { zones: ["wave_pool", "hot_tubs", "cabanas", "shark_reef", "beach_club", "beach_rave"] },
  },
  steve_at_the_reef: {
    id: "steve_at_the_reef",
    label: "Steve at the Reef",
    narrative: "Steve Harvey narrates the shark tunnel like Family Feud. Survey says… apex predator!",
    requires: { zones: ["shark_reef"], flags: ["steve_reef_rumor"] },
  },
};

export function defaultPoolComplexState(overrides = {}) {
  return {
    visitedZones: overrides.visitedZones ?? [],
    actionLog: overrides.actionLog ?? [],
    unlockedEvents: overrides.unlockedEvents ?? [],
    flags: overrides.flags ?? {},
    waveWins: overrides.waveWins ?? 0,
    ringTossWins: overrides.ringTossWins ?? 0,
    sharkPhotos: overrides.sharkPhotos ?? [],
    hotTubSoaks: overrides.hotTubSoaks ?? 0,
    waveTarget: overrides.waveTarget ?? null,
    raveMoves: overrides.raveMoves ?? [],
    raveStep: overrides.raveStep ?? 0,
    ...overrides,
  };
}

export function ensurePoolComplex(session) {
  if (!session.poolComplex) {
    session.poolComplex = defaultPoolComplexState();
  }
  const defaults = defaultPoolComplexState();
  const pc = session.poolComplex;
  for (const key of Object.keys(defaults)) {
    if (pc[key] === undefined) pc[key] = defaults[key];
  }
  if (!pc.flags) pc.flags = {};
  return pc;
}

export function attachPoolComplexToSession(session, data) {
  if (data.poolComplex) {
    session.poolComplex = { ...defaultPoolComplexState(), ...data.poolComplex };
  } else {
    ensurePoolComplex(session);
  }
}

function hasAll(haystack, needles) {
  return needles.every((n) => haystack.includes(n));
}

function eventRequirementsMet(session, pc, event) {
  const req = event.requires;
  if (req.zones && !hasAll(pc.visitedZones, req.zones)) return false;
  if (req.sharkPhotos && pc.sharkPhotos.length < req.sharkPhotos) return false;
  if (req.flags && !req.flags.every((f) => pc.flags[f])) return false;
  if (req.netPositive && !isNetPositive(session)) return false;
  return true;
}

function tryUnlockEvents(session) {
  const pc = ensurePoolComplex(session);
  const unlocked = [];
  for (const event of Object.values(POOL_EVENTS)) {
    if (pc.unlockedEvents.includes(event.id)) continue;
    if (!eventRequirementsMet(session, pc, event)) continue;
    pc.unlockedEvents.push(event.id);
    pc.actionLog.push(event.narrative);
    unlocked.push(event);
  }
  return unlocked;
}

function visitZone(session, zoneId) {
  const pc = ensurePoolComplex(session);
  if (!pc.visitedZones.includes(zoneId)) {
    pc.visitedZones.push(zoneId);
  }
  return tryUnlockEvents(session);
}

function finishResult(session, message, unlocked = []) {
  let msg = message;
  if (unlocked.length) {
    msg += `\n\n✦ Unlocked: ${unlocked.map((e) => e.label).join(", ")}`;
  }
  return { ok: true, message: msg, unlock: unlocked[0]?.id };
}

/** @returns {PoolResult} */
export function enterZone(session, zoneId) {
  const zone = POOL_ZONES[zoneId];
  if (!zone) return { ok: false, message: "That's a maintenance closet." };
  const unlocked = visitZone(session, zoneId);
  return finishResult(session, `${zone.label}\n${zone.description}`, unlocked);
}

/** @returns {PoolResult} */
export function playCatchWave(session, timingIndex) {
  const pc = ensurePoolComplex(session);
  visitZone(session, "wave_pool");
  if (pc.waveTarget === null) {
    pc.waveTarget = secureRandomInt(0, WAVE_TIMING.length - 1);
  }
  const pick = WAVE_TIMING[timingIndex];
  if (!pick) return { ok: false, message: "Pick a timing." };
  const correct = timingIndex === pc.waveTarget;
  pc.waveTarget = secureRandomInt(0, WAVE_TIMING.length - 1);
  if (correct) {
    pc.waveWins += 1;
    const payout = 25;
    session.wallet.credit(payout, "pool", "Wave pool ride");
    const unlocked = tryUnlockEvents(session);
    return finishResult(
      session,
      `${pick} — PERFECT! The wave lifts you. +${fmtChips(payout)}.\nChlorinated glory.`,
      unlocked,
    );
  }
  return { ok: true, message: `${pick} — mistimed. The wave eats you. Try again.` };
}

/** @returns {PoolResult} */
export function playRingToss(session, bet, ringIndex) {
  const pc = ensurePoolComplex(session);
  visitZone(session, "wave_pool");
  if (bet < 10) return { ok: false, message: "Minimum $10 ring toss." };
  if (!session.wallet.debit(bet, "pool", "Ring toss")) {
    return { ok: false, message: `Need ${fmtChips(bet)} for rings.` };
  }
  const rings = ["Inner tube", "Lifeguard tower", "Cabana post"];
  const pick = rings[ringIndex];
  if (!pick) return { ok: false, message: "Pick a target." };
  const roll = secureRandomInt(1, 10);
  if (roll >= 8) {
    const payout = bet * 3;
    session.wallet.credit(payout, "pool", "Ring toss winner");
    pc.ringTossWins += 1;
    const unlocked = tryUnlockEvents(session);
    return finishResult(session, `${pick} — RINGER! +${fmtChips(payout)}.`, unlocked);
  }
  if (roll >= 5) {
    session.wallet.credit(bet, "pool", "Ring toss push");
    return { ok: true, message: `${pick} — close! Bet returned.` };
  }
  return { ok: true, message: `${pick} — splash. Ring sunk. House keeps ${fmtChips(bet)}.` };
}

/** @returns {PoolResult} */
export function photographShark(session, speciesId) {
  const pc = ensurePoolComplex(session);
  visitZone(session, "shark_reef");
  const species = SHARK_SPECIES[speciesId];
  if (!species) return { ok: false, message: "That species migrated." };
  if (pc.sharkPhotos.includes(speciesId)) {
    return { ok: true, message: `${species.label} already in your camera roll.` };
  }
  pc.sharkPhotos.push(speciesId);
  const unlocked = tryUnlockEvents(session);
  return finishResult(
    session,
    `📷 ${species.label} captured! (${pc.sharkPhotos.length}/5 species)`,
    unlocked,
  );
}

/** @returns {PoolResult} */
export function soakHotTub(session, choiceId) {
  visitZone(session, "hot_tubs");
  const pc = ensurePoolComplex(session);
  pc.hotTubSoaks += 1;
  const options = {
    gossip: {
      message: "Overheard: Steve Harvey was at the reef yesterday narrating sharks. Survey says… fish!",
      flag: "steve_reef_rumor",
    },
    relax: { message: "You soak until your chips feel lighter. They aren't. You are.", flag: null },
    challenge: { message: "A stranger bets you can't last ten minutes without checking sportsbook odds. You last four.", flag: null },
  };
  const opt = options[choiceId];
  if (!opt) return { ok: false, message: "Pick a soak option." };
  if (opt.flag) pc.flags[opt.flag] = true;
  const unlocked = tryUnlockEvents(session);
  return finishResult(session, opt.message, unlocked);
}

/** @returns {PoolResult} */
export function bookCabana(session) {
  visitZone(session, "cabanas");
  const pc = ensurePoolComplex(session);
  if (pc.flags.cabana_booked) {
    return { ok: true, message: "Your cabana awaits. Bottle service on speed dial." };
  }
  const cost = 200;
  if (!session.wallet.debit(cost, "pool", "Cabana rental")) {
    return { ok: false, message: `Cabana requires ${fmtChips(cost)} or a Pearl+ comp.` };
  }
  pc.flags.cabana_booked = true;
  const unlocked = tryUnlockEvents(session);
  return finishResult(session, `Cabana secured for ${fmtChips(cost)}. Privacy, shade, and judgment-free naps.`, unlocked);
}

/** @returns {PoolResult} */
export function cabanaService(session, serviceId) {
  const pc = ensurePoolComplex(session);
  if (!pc.flags.cabana_booked) {
    return { ok: false, message: "Book a cabana first." };
  }
  const services = {
    bottle: { price: 85, message: "Champagne arrives on ice. The cabana approves." },
    nap: { price: 0, message: "You nap through the afternoon heat. Smart money." },
    people_watch: { price: 0, message: "You people-watch the wave pool. Entertainment is free." },
  };
  const svc = services[serviceId];
  if (!svc) return { ok: false, message: "Unknown service." };
  if (svc.price && !session.wallet.debit(svc.price, "pool", "Cabana service")) {
    return { ok: false, message: `Need ${fmtChips(svc.price)}.` };
  }
  if (serviceId === "bottle") {
    recordConsumption(session, POOL_CONSUMABLE_IDS.bottle, { source: "pool_cabana" });
  }
  return { ok: true, message: svc.message };
}

/** @returns {PoolResult} */
export function enterBeachClub(session) {
  visitZone(session, "beach_club");
  const pc = ensurePoolComplex(session);
  if (pc.flags.beach_club_pass) {
    const unlocked = tryUnlockEvents(session);
    return finishResult(session, "Welcome back to the beach club. The rope remembers you.", unlocked);
  }
  const cover = 75;
  if (!session.wallet.debit(cover, "pool", "Beach club cover")) {
    return { ok: false, message: `Cover charge: ${fmtChips(cover)}. Confidence sold separately.` };
  }
  pc.flags.beach_club_pass = true;
  const unlocked = tryUnlockEvents(session);
  return finishResult(
    session,
    `Cover paid (${fmtChips(cover)}). European-style deck. 21+ enforced. Sunscreen optional.`,
    unlocked,
  );
}

/** @returns {PoolResult} */
export function beachClubAction(session, actionId) {
  if (!ensurePoolComplex(session).flags.beach_club_pass) {
    return enterBeachClub(session);
  }
  const actions = {
    bar: { price: 18, message: "Frozen cocktail that costs more than your first apartment." },
    sun_deck: { price: 0, message: "You claim a lounger. The Vegas sun files a tax return on your skin." },
    vip_rope: { price: 50, message: "VIP section. Fewer people, same bad decisions." },
  };
  const act = actions[actionId];
  if (!act) return { ok: false, message: "Pick an action." };
  if (act.price && !session.wallet.debit(act.price, "pool", "Beach club")) {
    return { ok: false, message: `Need ${fmtChips(act.price)}.` };
  }
  if (actionId === "bar") {
    recordConsumption(session, POOL_CONSUMABLE_IDS.bar, { source: "pool_beach_club" });
  }
  return { ok: true, message: act.message };
}

/** @returns {PoolResult} */
export function startRaveDance(session) {
  visitZone(session, "beach_rave");
  const pc = ensurePoolComplex(session);
  pc.raveMoves = [secureRandomInt(0, 2), secureRandomInt(0, 2), secureRandomInt(0, 2)];
  pc.raveStep = 0;
  return { ok: true, message: "The DJ drops bass. Match three moves: Fist pump, Shuffling, Glow spin." };
}

/** @returns {PoolResult} */
export function submitRaveMove(session, moveIndex) {
  const pc = ensurePoolComplex(session);
  if (!pc.raveMoves.length) return startRaveDance(session);
  const moves = ["Fist pump", "Shuffling", "Glow spin"];
  if (pc.raveMoves[pc.raveStep] !== moveIndex) {
    pc.raveMoves = [];
    pc.raveStep = 0;
    return { ok: true, message: `${moves[moveIndex] ?? "?"} — off beat. Crowd wobbles. Try again.` };
  }
  pc.raveStep += 1;
  if (pc.raveStep >= pc.raveMoves.length) {
    pc.flags.rave_danced = true;
    pc.raveMoves = [];
    pc.raveStep = 0;
    const payout = 40;
    session.wallet.credit(payout, "pool", "Beach rave");
    const unlocked = tryUnlockEvents(session);
    return finishResult(session, `Sequence hit! The wave pool glows. +${fmtChips(payout)}.`, unlocked);
  }
  return { ok: true, message: `${moves[moveIndex]} — on beat! ${pc.raveStep + 1}/3…` };
}

export function getPoolSummary(session) {
  const pc = ensurePoolComplex(session);
  const parts = [];
  if (pc.visitedZones.length) parts.push(`${pc.visitedZones.length}/6 zones`);
  if (pc.sharkPhotos.length) parts.push(`${pc.sharkPhotos.length}/5 reef photos`);
  if (pc.unlockedEvents.length) parts.push(`${pc.unlockedEvents.length} event(s)`);
  if (pc.flags.cabana_booked) parts.push("cabana booked");
  return parts.length ? parts.join(" · ") : "Eleven acres await.";
}

export function getUnlockedPoolEvents(session) {
  const pc = ensurePoolComplex(session);
  return pc.unlockedEvents.map((id) => POOL_EVENTS[id]).filter(Boolean);
}

File: mandalay_bay/pool_complex.py
"""Mandalay Bay Pool Complex expansion — zones, minigames, unlockable events."""

from __future__ import annotations

import random
from dataclasses import dataclass, field

from mandalay_bay.hotel import fmt_chips, is_net_positive
from mandalay_bay.session import PlayerSession

POOL_ZONES = {
    "wave_pool": {"label": "Wave Pool", "description": "Eleven acres of artificial surf."},
    "hot_tubs": {"label": "Hot Tubs", "description": "Steaming circles of gossip and chlorine."},
    "cabanas": {"label": "Private Cabanas", "description": "Rope-off luxury and bottle service."},
    "shark_reef": {"label": "Shark Reef Aquarium", "description": "Sand tiger sharks behind acrylic."},
    "beach_club": {"label": "Topless Beach Club", "description": "21+ European-style pool deck."},
    "beach_rave": {"label": "Beach Rave", "description": "Neon after dark by the wave pool."},
}

SHARK_SPECIES = {
    "sand_tiger": {"label": "Sand tiger shark"},
    "green_sea_turtle": {"label": "Green sea turtle"},
    "sawfish": {"label": "Largetooth sawfish"},
    "golden_ray": {"label": "Golden cownose ray"},
    "jellyfish": {"label": "Pacific sea nettle"},
}

WAVE_TIMING = ["Jump early", "Ride the crest", "Bail late"]
RAVE_MOVES = ["Fist pump", "Shuffling", "Glow spin"]

POOL_EVENTS = {
    "first_splash": {
        "label": "First Splash",
        "narrative": "The wave catches you perfectly. Chlorinated triumph.",
        "requires": {"zones": ["wave_pool"]},
    },
    "steam_and_chips": {
        "label": "Steam & Chips",
        "narrative": "Hot tub gossip confirms your blackjack run.",
        "requires": {"zones": ["hot_tubs"], "net_positive": True},
    },
    "cabana_king": {
        "label": "Cabana King",
        "narrative": "Your cabana. Your rules.",
        "requires": {"flags": ["cabana_booked"]},
    },
    "reef_photographer": {
        "label": "Reef Photographer",
        "narrative": "Five species captured.",
        "requires": {"shark_photos": 5},
    },
    "beach_club_initiate": {
        "label": "Beach Club Initiate",
        "narrative": "You survived the rope line.",
        "requires": {"zones": ["beach_club"]},
    },
    "rave_til_dawn": {
        "label": "Rave Til Dawn",
        "narrative": "Glow sticks and bass drops until dawn.",
        "requires": {"zones": ["beach_rave"], "flags": ["rave_danced"]},
    },
    "eleven_acres": {
        "label": "Eleven Acres Conquered",
        "narrative": "Every zone in the complex.",
        "requires": {"zones": list(POOL_ZONES.keys())},
    },
    "steve_at_the_reef": {
        "label": "Steve at the Reef",
        "narrative": "Steve Harvey narrates the shark tunnel. Survey says… apex predator!",
        "requires": {"zones": ["shark_reef"], "flags": ["steve_reef_rumor"]},
    },
}


@dataclass
class PoolComplexState:
    visited_zones: list[str] = field(default_factory=list)
    action_log: list[str] = field(default_factory=list)
    unlocked_events: list[str] = field(default_factory=list)
    flags: dict[str, bool] = field(default_factory=dict)
    wave_wins: int = 0
    ring_toss_wins: int = 0
    shark_photos: list[str] = field(default_factory=list)
    hot_tub_soaks: int = 0
    wave_target: int | None = None
    rave_moves: list[int] = field(default_factory=list)
    rave_step: int = 0


@dataclass
class PoolResult:
    ok: bool
    message: str
    unlock: str | None = None


def default_pool_complex_state() -> PoolComplexState:
    return PoolComplexState()


def ensure_pool_complex(session: PlayerSession) -> PoolComplexState:
    if not hasattr(session, "pool_complex") or session.pool_complex is None:
        session.pool_complex = default_pool_complex_state()
    return session.pool_complex


def _has_all(haystack: list[str], needles: list[str]) -> bool:
    return all(n in haystack for n in needles)


def _requirements_met(session: PlayerSession, pc: PoolComplexState, event: dict) -> bool:
    req = event["requires"]
    if "zones" in req and not _has_all(pc.visited_zones, req["zones"]):
        return False
    if "shark_photos" in req and len(pc.shark_photos) < req["shark_photos"]:
        return False
    if "flags" in req and not all(pc.flags.get(f) for f in req["flags"]):
        return False
    if req.get("net_positive") and not is_net_positive(session):
        return False
    return True


def _try_unlock_events(session: PlayerSession) -> list[dict]:
    pc = ensure_pool_complex(session)
    unlocked = []
    for event_id, event in POOL_EVENTS.items():
        if event_id in pc.unlocked_events:
            continue
        if not _requirements_met(session, pc, event):
            continue
        pc.unlocked_events.append(event_id)
        pc.action_log.append(event["narrative"])
        unlocked.append(event)
    return unlocked


def _visit_zone(session: PlayerSession, zone_id: str) -> list[dict]:
    pc = ensure_pool_complex(session)
    if zone_id not in pc.visited_zones:
        pc.visited_zones.append(zone_id)
    return _try_unlock_events(session)


def _finish(message: str, unlocked: list[dict]) -> PoolResult:
    if unlocked:
        labels = ", ".join(e["label"] for e in unlocked)
        message += f"\n\nUnlocked: {labels}"
    return PoolResult(True, message)


def enter_zone(session: PlayerSession, zone_id: str) -> PoolResult:
    zone = POOL_ZONES.get(zone_id)
    if not zone:
        return PoolResult(False, "That's a maintenance closet.")
    unlocked = _visit_zone(session, zone_id)
    return _finish(f"{zone['label']}\n{zone['description']}", unlocked)


def play_catch_wave(session: PlayerSession, timing_index: int) -> PoolResult:
    pc = ensure_pool_complex(session)
    _visit_zone(session, "wave_pool")
    if pc.wave_target is None:
        pc.wave_target = random.randint(0, len(WAVE_TIMING) - 1)
    if timing_index < 0 or timing_index >= len(WAVE_TIMING):
        return PoolResult(False, "Pick a timing.")
    pick = WAVE_TIMING[timing_index]
    correct = timing_index == pc.wave_target
    pc.wave_target = random.randint(0, len(WAVE_TIMING) - 1)
    if correct:
        pc.wave_wins += 1
        payout = 25
        session.wallet.credit(payout, "pool", "Wave pool ride")
        unlocked = _try_unlock_events(session)
        return _finish(f"{pick} — PERFECT! +{fmt_chips(payout)}.", unlocked)
    return PoolResult(True, f"{pick} — mistimed. Try again.")


def play_ring_toss(session: PlayerSession, bet: int, ring_index: int) -> PoolResult:
    pc = ensure_pool_complex(session)
    _visit_zone(session, "wave_pool")
    if bet < 10:
        return PoolResult(False, "Minimum $10 ring toss.")
    if not session.wallet.debit(bet, "pool", "Ring toss"):
        return PoolResult(False, f"Need {fmt_chips(bet)}.")
    rings = ["Inner tube", "Lifeguard tower", "Cabana post"]
    if ring_index < 0 or ring_index >= len(rings):
        return PoolResult(False, "Pick a target.")
    pick = rings[ring_index]
    roll = random.randint(1, 10)
    if roll >= 8:
        payout = bet * 3
        session.wallet.credit(payout, "pool", "Ring toss winner")
        pc.ring_toss_wins += 1
        unlocked = _try_unlock_events(session)
        return _finish(f"{pick} — RINGER! +{fmt_chips(payout)}.", unlocked)
    if roll >= 5:
        session.wallet.credit(bet, "pool", "Ring toss push")
        return PoolResult(True, f"{pick} — close! Bet returned.")
    return PoolResult(True, f"{pick} — splash. House keeps {fmt_chips(bet)}.")


def photograph_shark(session: PlayerSession, species_id: str) -> PoolResult:
    pc = ensure_pool_complex(session)
    _visit_zone(session, "shark_reef")
    species = SHARK_SPECIES.get(species_id)
    if not species:
        return PoolResult(False, "That species migrated.")
    if species_id in pc.shark_photos:
        return PoolResult(True, f"{species['label']} already photographed.")
    pc.shark_photos.append(species_id)
    unlocked = _try_unlock_events(session)
    return _finish(f"{species['label']} captured! ({len(pc.shark_photos)}/5)", unlocked)


def soak_hot_tub(session: PlayerSession, choice_id: str) -> PoolResult:
    _visit_zone(session, "hot_tubs")
    pc = ensure_pool_complex(session)
    pc.hot_tub_soaks += 1
    options = {
        "gossip": ("Steve Harvey was at the reef narrating sharks. Survey says… fish!", "steve_reef_rumor"),
        "relax": ("You soak until your chips feel lighter.", None),
        "challenge": ("You last four minutes without checking odds.", None),
    }
    opt = options.get(choice_id)
    if not opt:
        return PoolResult(False, "Pick a soak option.")
    message, flag = opt
    if flag:
        pc.flags[flag] = True
    unlocked = _try_unlock_events(session)
    return _finish(message, unlocked)


def book_cabana(session: PlayerSession) -> PoolResult:
    _visit_zone(session, "cabanas")
    pc = ensure_pool_complex(session)
    if pc.flags.get("cabana_booked"):
        return PoolResult(True, "Your cabana awaits.")
    cost = 200
    if not session.wallet.debit(cost, "pool", "Cabana rental"):
        return PoolResult(False, f"Cabana requires {fmt_chips(cost)}.")
    pc.flags["cabana_booked"] = True
    unlocked = _try_unlock_events(session)
    return _finish(f"Cabana secured for {fmt_chips(cost)}.", unlocked)


def cabana_service(session: PlayerSession, service_id: str) -> PoolResult:
    pc = ensure_pool_complex(session)
    if not pc.flags.get("cabana_booked"):
        return PoolResult(False, "Book a cabana first.")
    services = {
        "bottle": (85, "Champagne arrives on ice."),
        "nap": (0, "You nap through the afternoon heat."),
        "people_watch": (0, "People-watching the wave pool."),
    }
    svc = services.get(service_id)
    if not svc:
        return PoolResult(False, "Unknown service.")
    price, message = svc
    if price and not session.wallet.debit(price, "pool", "Cabana service"):
        return PoolResult(False, f"Need {fmt_chips(price)}.")
    return PoolResult(True, message)


def enter_beach_club(session: PlayerSession) -> PoolResult:
    _visit_zone(session, "beach_club")
    pc = ensure_pool_complex(session)
    if pc.flags.get("beach_club_pass"):
        unlocked = _try_unlock_events(session)
        return _finish("Welcome back to the beach club.", unlocked)
    cover = 75
    if not session.wallet.debit(cover, "pool", "Beach club cover"):
        return PoolResult(False, f"Cover charge: {fmt_chips(cover)}.")
    pc.flags["beach_club_pass"] = True
    unlocked = _try_unlock_events(session)
    return _finish(f"Cover paid ({fmt_chips(cover)}). 21+ enforced.", unlocked)


def beach_club_action(session: PlayerSession, action_id: str) -> PoolResult:
    pc = ensure_pool_complex(session)
    if not pc.flags.get("beach_club_pass"):
        return enter_beach_club(session)
    actions = {
        "bar": (18, "Frozen cocktail acquired."),
        "sun_deck": (0, "You claim a lounger."),
        "vip_rope": (50, "VIP section accessed."),
    }
    act = actions.get(action_id)
    if not act:
        return PoolResult(False, "Pick an action.")
    price, message = act
    if price and not session.wallet.debit(price, "pool", "Beach club"):
        return PoolResult(False, f"Need {fmt_chips(price)}.")
    return PoolResult(True, message)


def start_rave_dance(session: PlayerSession) -> PoolResult:
    _visit_zone(session, "beach_rave")
    pc = ensure_pool_complex(session)
    pc.rave_moves = [random.randint(0, 2) for _ in range(3)]
    pc.rave_step = 0
    return PoolResult(True, "Match three moves: Fist pump, Shuffling, Glow spin.")


def submit_rave_move(session: PlayerSession, move_index: int) -> PoolResult:
    pc = ensure_pool_complex(session)
    if not pc.rave_moves:
        return start_rave_dance(session)
    if move_index < 0 or move_index > 2:
        return PoolResult(False, "Pick a move.")
    if pc.rave_moves[pc.rave_step] != move_index:
        pc.rave_moves.clear()
        pc.rave_step = 0
        return PoolResult(True, f"{RAVE_MOVES[move_index]} — off beat. Try again.")
    pc.rave_step += 1
    if pc.rave_step >= len(pc.rave_moves):
        pc.flags["rave_danced"] = True
        pc.rave_moves.clear()
        pc.rave_step = 0
        payout = 40
        session.wallet.credit(payout, "pool", "Beach rave")
        unlocked = _try_unlock_events(session)
        return _finish(f"Sequence hit! +{fmt_chips(payout)}.", unlocked)
    return PoolResult(True, f"{RAVE_MOVES[move_index]} — on beat! {pc.rave_step + 1}/3…")


def get_pool_summary(session: PlayerSession) -> str:
    pc = ensure_pool_complex(session)
    parts = []
    if pc.visited_zones:
        parts.append(f"{len(pc.visited_zones)}/6 zones")
    if pc.shark_photos:
        parts.append(f"{len(pc.shark_photos)}/5 reef photos")
    if pc.unlocked_events:
        parts.append(f"{len(pc.unlocked_events)} event(s)")
    if pc.flags.get("cabana_booked"):
        parts.append("cabana booked")
    return " · ".join(parts) if parts else "Eleven acres await."
